
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Command-line utilities &#8212; grstools 0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="File formats" href="formats.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">grstools</a></h1>










<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="formats.html">File formats</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Command-line utilities</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#grs-compute"><code class="docutils literal"><span class="pre">grs-compute</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#grs-create"><code class="docutils literal"><span class="pre">grs-create</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#grs-mr"><code class="docutils literal"><span class="pre">grs-mr</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#grs-utils"><code class="docutils literal"><span class="pre">grs-utils</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#grs-evaluate"><code class="docutils literal"><span class="pre">grs-evaluate</span></code></a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="formats.html" title="previous chapter">File formats</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="command-line-utilities">
<h1>Command-line utilities<a class="headerlink" href="#command-line-utilities" title="Permalink to this headline">¶</a></h1>
<div class="section" id="grs-compute">
<span id="id1"></span><h2><code class="docutils literal"><span class="pre">grs-compute</span></code><a class="headerlink" href="#grs-compute" title="Permalink to this headline">¶</a></h2>
<p>Computes the GRS using a ‘.grs’ file and individual-level genotypes.</p>
<p>The GRS are computed using the following formula:</p>
<div class="math">
\[GRS_i = \sum_{j=0}^{m} \beta_j \cdot X_{ij}\]</div>
<p>Where <span class="math">\(GRS_i\)</span> is the genetic risk score for the <cite>i</cite><sup>th</sup> sample, j
indexes the genetic variants included in the score, <span class="math">\(\beta_j\)</span> is the
coefficient estimated from the summary statistics and <span class="math">\(X_{ij}\)</span> is the
number of effect alleles carried by individual <cite>i</cite> at variant <cite>j</cite>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If the coefficient (<span class="math">\(\beta\)</span>) is negative, it’s additive inverse is
used instead (<span class="math">\(-\beta\)</span>) and the effect and reference alleles are
swapped accordingly. This ensures that contributions to the GRS are always
positive.</p>
</div>
<p>If the IMPUTE2 format is used to hold the genotype information, the variants
are further weighted by the INFO score as a form of genotype confidence
weighting:</p>
<div class="math">
\[GRS_i = \sum_{j=0}^{m} \beta_j \cdot X_{ij}  \cdot info_j\]</div>
<div class="admonition-todo admonition" id="index-0">
<p class="first admonition-title">Todo</p>
<p class="last">Eventually, we will weight variant as soon as there is some genotype
quality metric (e.g. in a VCF file).</p>
</div>
<div class="highlight-none"><div class="highlight"><pre><span></span>usage: grs-compute [-h] [--genotypes GENOTYPES]
                   [--genotypes-format GENOTYPES_FORMAT]
                   [--genotypes-kwargs GENOTYPES_KWARGS] --grs GRS [--out OUT]
                   [--ignore-genotype-quality] [--skip-strand-check]
                   [--exclude-strand-ambiguous] [--reference REFERENCE]
                   [--debug]

Compute the risk score.

optional arguments:
  -h, --help            show this help message and exit
  --genotypes GENOTYPES
                        File containing genotype data.
  --genotypes-format GENOTYPES_FORMAT
                        File format of the genotypes in the reference
                        (default: plink).
  --genotypes-kwargs GENOTYPES_KWARGS
                        Keyword arguments to pass to the genotypes container.
                        A string of the following format is expected:
                        &#39;key1=value1,key2=value2,...It is also possible to
                        prefix the values by &#39;int:&#39; or &#39;float:&#39; to cast the
                        them before passing them to the constructor.
  --grs GRS             File describing the variants in the GRS. The expected
                        columns are &#39;name&#39;, &#39;beta&#39;, &#39;reference&#39;, &#39;risk&#39;.
  --out OUT             Filename for the computed GRS (default:
                        computed_grs.csv).
  --ignore-genotype-quality
                        For imputed variants, if this flag is set, the
                        variants will not be weighted with respect to their
                        quality score. By default, the weights are used if
                        available.
  --skip-strand-check   Skips all the strand validation checks and assume that
                        the strand in the genotype and GRS files are the same.
                        This option should only be used if the variants were
                        manually curated. This option is also assumed if no
                        --reference is provided (with a warning).
  --exclude-strand-ambiguous
                        Exclude all strand ambiguous variants.
  --reference REFERENCE
                        Plink prefix for the reference genotypes. This is used
                        to find tags when needed or to compute the MAF.
  --debug
</pre></div>
</div>
</div>
<div class="section" id="grs-create">
<span id="id2"></span><h2><code class="docutils literal"><span class="pre">grs-create</span></code><a class="headerlink" href="#grs-create" title="Permalink to this headline">¶</a></h2>
<p>Creating a GRS consists of selecting variants that will be considered jointly
when the GRS will be computed using genotype data.</p>
<p>This script requires a <code class="docutils literal"><span class="pre">.grs</span></code> file and the selection is based on the p-value
sorting and thresholding algorithm.</p>
<p>This method consists of the following steps:</p>
<ol class="arabic simple">
<li>Rank the variants by increasing p-value.</li>
<li>Select the top variant, include it in the GRS and exclude all variants in LD
at a pre-defined threshold (<code class="docutils literal"><span class="pre">--ld-threshold</span></code>).</li>
<li>Repeat 2. until the p-value threshold is reached (<code class="docutils literal"><span class="pre">--p-threshold</span></code>).</li>
</ol>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If a <code class="docutils literal"><span class="pre">maf</span></code> column is present in the summary statistics association file,
all variants that are too rare will be automaticall excluded when reading
the summary statistics file. If it is not available, the MAF will be
computed from the genotype data to allow filtering, which is more
computationally intensive.</p>
</div>
<p>For now, a plink binary file is required for the reference. A
population-specific reference panel like the 1000 genomes phase 3 should be
used.</p>
<p>This script generates a ‘grs’ file that is suitable for computing the GRS using
<a class="reference internal" href="#grs-compute"><span class="std std-ref">grs-compute</span></a>.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>usage: grs-create [-h] [--p-threshold P_THRESHOLD] [--target-n TARGET_N]
                  [--maf-threshold MAF_THRESHOLD]
                  [--ld-threshold LD_THRESHOLD]
                  [--ld-window-size LD_WINDOW_SIZE] [--region REGION]
                  [--exclude-region EXCLUDE_REGION] [--keep-ambiguous-alleles]
                  --summary SUMMARY --reference REFERENCE [--output OUTPUT]
                  [--debug]

optional arguments:
  -h, --help            show this help message and exit
  --p-threshold P_THRESHOLD
                        P-value threshold for inclusion in the GRS (default:
                        5e-8).
  --target-n TARGET_N   Target number of variants to include in the GRS.
  --maf-threshold MAF_THRESHOLD
                        Minimum MAF to allow inclusion in the GRS (default
                        0.05).
  --ld-threshold LD_THRESHOLD
                        LD threshold for the clumping step. All variants in LD
                        with variants in the GRS are excluded iteratively
                        (default 0.05).
  --ld-window-size LD_WINDOW_SIZE
                        Size of the LD window used to find correlated
                        variants. Making this window smaller will make the
                        execution faster but increases the chance of missing
                        correlated variants (default 500kb).
  --region REGION       Only consider variants located within a genomic
                        region. The expected format is &#39;chrCHR:START-END&#39;. For
                        example: &#39;chr1:12345-22345&#39;.
  --exclude-region EXCLUDE_REGION
                        Only consider variants located OUTSIDE a genomic
                        region. The expected format is &#39;chrCHR:START-END&#39;. For
                        example: &#39;chr1:12345-22345&#39;.
  --keep-ambiguous-alleles
                        Do not filter out ambiguous alleles (e.g. G/C or A/T)
  --summary SUMMARY     Path to the summary statistics files. Required columns
                        are &#39;name&#39;, &#39;chrom&#39;, &#39;pos&#39;, &#39;p-value&#39;, &#39;effect&#39;,
                        &#39;reference&#39; and &#39;risk&#39;.
  --reference REFERENCE
                        Path the binary plink file containing reference
                        genotypes. These genotypes will be used for LD
                        clumping.
  --output OUTPUT, -o OUTPUT
                        Output filename (default: selected.grs).
  --debug
</pre></div>
</div>
</div>
<div class="section" id="grs-mr">
<span id="id3"></span><h2><code class="docutils literal"><span class="pre">grs-mr</span></code><a class="headerlink" href="#grs-mr" title="Permalink to this headline">¶</a></h2>
<p>Mendelian randomization is a technique where genetic variants (or a genetic
risk score (G)) is used to estimate the effect of an exposure (X) on an outcome
(Y). Here, the estimation is done using the ratio method:</p>
<div class="math">
\[\beta = \frac{\beta_{Y \sim G}}{\beta_{X \sim G}}\]</div>
<p>In the current implementation, bootstrap resampling is used to estimate the
standard error which can be used for further inference.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>usage: grs-mr [-h] [--grs-filename GRS_FILENAME] [--exposure EXPOSURE]
              [--outcome OUTCOME] [--exposure-type EXPOSURE_TYPE]
              [--outcome-type OUTCOME_TYPE]
              [--phenotypes-filename PHENOTYPES_FILENAME]
              [--phenotypes-sample-column PHENOTYPES_SAMPLE_COLUMN]
              [--phenotypes-separator PHENOTYPES_SEPARATOR]
              [--phenotype PHENOTYPE]

Estimate the effect of an exposure on an outcome using a GRS with an effect on
the exposure. Estimates are done using the ratio method.

optional arguments:
  -h, --help            show this help message and exit
  --grs-filename GRS_FILENAME
  --exposure EXPOSURE
  --outcome OUTCOME
  --exposure-type EXPOSURE_TYPE
                        Either continuous or discrete.
  --outcome-type OUTCOME_TYPE
                        Either continuous or discrete.
  --phenotypes-filename PHENOTYPES_FILENAME
  --phenotypes-sample-column PHENOTYPES_SAMPLE_COLUMN
  --phenotypes-separator PHENOTYPES_SEPARATOR
  --phenotype PHENOTYPE
</pre></div>
</div>
</div>
<div class="section" id="grs-utils">
<span id="id4"></span><h2><code class="docutils literal"><span class="pre">grs-utils</span></code><a class="headerlink" href="#grs-utils" title="Permalink to this headline">¶</a></h2>
<p>This tool contains multiple features to manipulate GRS after they have been
computed. Most of these tools either generate plots (as png files) or modified
versions of the computed GRS. The latter file format is simply a CSV with a
‘sample’ and a ‘grs’ column.</p>
<div class="section" id="histogram">
<h3>histogram<a class="headerlink" href="#histogram" title="Permalink to this headline">¶</a></h3>
<p>The histogram sub-command generates a histogram of all the GRS values. Here is
an example:</p>
<img alt="_images/histogram.png" src="_images/histogram.png" />
</div>
<div class="section" id="quantiles">
<h3>quantiles<a class="headerlink" href="#quantiles" title="Permalink to this headline">¶</a></h3>
<p>This function is used to dichotomize the GRS. The parameters are:
- <code class="docutils literal"><span class="pre">-k</span></code> The number of quantiles to take to form a group.
- <code class="docutils literal"><span class="pre">-q</span></code> The total number of quantiles.</p>
<p>Here are various examples of dichotomizations:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1"># Create two groups with respect to the median.</span>
<span class="n">grs</span><span class="o">-</span><span class="n">utils</span> <span class="n">quantiles</span> <span class="n">my_grs</span><span class="o">.</span><span class="n">grs</span> <span class="o">-</span><span class="n">k</span> <span class="mi">1</span> <span class="o">-</span><span class="n">q</span> <span class="mi">2</span>

<span class="c1"># 1st vs 5th quintiles.</span>
<span class="n">grs</span><span class="o">-</span><span class="n">utils</span> <span class="n">quantiles</span> <span class="n">my_grs</span><span class="o">.</span><span class="n">grs</span> <span class="o">-</span><span class="n">k</span> <span class="mi">1</span> <span class="o">-</span><span class="n">q</span> <span class="mi">5</span>

<span class="c1"># 1st vs 5th quintiles, but keeping all the samples (will be set to NA).</span>
<span class="n">grs</span><span class="o">-</span><span class="n">utils</span> <span class="n">quantiles</span> <span class="n">my_grs</span><span class="o">.</span><span class="n">grs</span> <span class="o">-</span><span class="n">k</span> <span class="mi">1</span> <span class="o">-</span><span class="n">q</span> <span class="mi">5</span> <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">unclassified</span>
</pre></div>
</div>
<p>The output file will contain two columns and a header: <code class="docutils literal"><span class="pre">sample</span></code> and
<code class="docutils literal"><span class="pre">group</span></code>. The group ‘0’ contains the individuals with lower values of the GRS
and the group ‘1’ contains individuals with high values of the GRS.</p>
</div>
<div class="section" id="standardize">
<h3>standardize<a class="headerlink" href="#standardize" title="Permalink to this headline">¶</a></h3>
<p>This function standardizes a GRS:</p>
<div class="math">
\[GRS' = \frac{GRS - \bar{GRS}}{s}\]</div>
<p>This function is useful when different GRS are to be compared. For example, is
linear regression is used to assess the effect of a GRS on a given trait (Y),
then the coefficient can be interpreted as the number of units increase in Y
for a 1 s.d. increase in the GRS.</p>
</div>
<div class="section" id="correlation">
<h3>correlation<a class="headerlink" href="#correlation" title="Permalink to this headline">¶</a></h3>
<p>This utility takes two computed GRS files as input and plots the correlation
between the GRS. This can be useful to test the variability of changing
different hyperparameters (e.g. p-value threshold).</p>
<p>This script opens the plot in the interactive matplotlib viewer. From there, it
can manipulated and saved to a file.</p>
<p>Here is an example:</p>
<img alt="_images/correlation.png" src="_images/correlation.png" />
<p>On the plot, the linear regression fit is displayed as well as the identity
line to quickly assess the agreement between GRS.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>usage: grs-utils [-h]
                 {histogram,quantiles,standardize,correlation,beta-plot} ...

Utilities to manipulate computed GRS.

positional arguments:
  {histogram,quantiles,standardize,correlation,beta-plot}
    histogram           Plot the histogram of the computed GRS.
    quantiles           Dichotomize the GRS using quantiles. Takes two
                        parameters: k and q where q is the number of quantiles
                        and k is the cutoff to be used for the discretization.
                        For example, if the 1st quintile is to be compared to
                        the 5th, use -q 5 -k 1. By default -k=1 and -q=2 which
                        means the median is used.
    standardize         Standardize the GRS (grs &lt;- (grs - mean) / std).
    correlation         Plot the correlation between two GRS.
    beta-plot           Compute beta coefficients from given genotypes data
                        and compare them with beta coefficients from the grs
                        file.

optional arguments:
  -h, --help            show this help message and exit
</pre></div>
</div>
</div>
</div>
<div class="section" id="grs-evaluate">
<span id="id5"></span><h2><code class="docutils literal"><span class="pre">grs-evaluate</span></code><a class="headerlink" href="#grs-evaluate" title="Permalink to this headline">¶</a></h2>
<p>After calculating a GRS, it is important to quantify and evaluate its
predictive performance. This tool provides multiple commands to easily run
common diagnostics and validation steps.</p>
<div class="section" id="regress">
<h3>regress<a class="headerlink" href="#regress" title="Permalink to this headline">¶</a></h3>
<p>The regress subcommand assesses the effect of the GRS as a continuous trait on
a given binary or continuous phenotype. In the case of a continuous outcome,
linear regression is used and a plot like the following is generated:</p>
<img alt="_images/evaluate-regress-continuous.png" src="_images/evaluate-regress-continuous.png" />
<p>In this plot, it is possible to quickly validate that the GRS indeed has an
effect on the trait of interest. It is also possible to estimate the fraction
of the variance explained by the GRS (corresponding to the <span class="math">\(R^2\)</span> value or
to evaluate the expected change in the phenotype attributable to a 1 unit
change in the GRS (<span class="math">\(\beta\)</span> coefficient and its 95% CI). Note that
standardizing the GRS first is useful to get an interpretation in units of GRS
s.d. and to make comparisons across different GRS possible.</p>
<p>If the outcome is discrete, then logistic regression is better suited to
evaluate the fit with the GRS. In this case, the generated plot looks as
follows:</p>
<img alt="_images/evaluate-regress-binary.png" src="_images/evaluate-regress-binary.png" />
<p>In this plot, the perspective is somewhat reversed, when compared with the
linear case. The GRS is now in the y axis and its distribution is shown in
cases and controls. The more general term “level” is used to describe the
discrete numerical values of the phenotype as it could be used for factor
variables.</p>
<p>The results for the logistic regression are shown on the plot. The OR
corresponds to the expected increase in odds for a one unit increase in the
GRS.</p>
</div>
<div class="section" id="dichotomize-plot">
<h3>dichotomize-plot<a class="headerlink" href="#dichotomize-plot" title="Permalink to this headline">¶</a></h3>
<p>This plot can be used to select an optimal quantile for dichotmization. The
tradeoff between the number of individuals classified and the GRS effect size
is highlighted. Note that selecting a local maxima in this plot can be a bad
idea because it increases the risk of overfitting. Traditional approaches such
as using a test set, using cross-validation or bootstrapping can be used to
select a more robust threshold.</p>
<p>The dichotmization is done by comparing the extremums of the distribution. For
example, at the 0.25 quantile, the first quartile (1/4) is compared to the last
quartile (4/4). All of the individuals in between can be used to form an
“intermediate” group or can be excluded from downstream analyses. This is why
the right axis represents the number of individuals that are classified in
either the “high” or “low” group for a given dichotmization.</p>
<img alt="_images/evaluate-dichotomize-plot.png" src="_images/evaluate-dichotomize-plot.png" />
</div>
<div class="section" id="roc">
<h3>ROC<a class="headerlink" href="#roc" title="Permalink to this headline">¶</a></h3>
<p>To evaluate the predictive accuracy of a GRS, it can be useful to look at its
ROC curve.</p>
<img alt="_images/evaluate-roc.png" src="_images/evaluate-roc.png" />
<div class="highlight-none"><div class="highlight"><pre><span></span>usage: grs-evaluate [-h] {regress,dichotomize-plot,roc} ...

Utilities to evaluate the performance of GRS.

positional arguments:
  {regress,dichotomize-plot,roc}
    regress             Regress the GRS on a discrete or continuous outcome.
    dichotomize-plot    A plot to help identify ideal dichotmizatin
                        parameters.
    roc                 Draw a ROC curve for a GRS (given a binary phenotype).

optional arguments:
  -h, --help            show this help message and exit
</pre></div>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2017, Marc-André Legault.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.1</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/utilities.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>